docs/contributing/muzzle.md

```diff
@@ -0,0 +1,32 @@
+# Muzzle
+
+Muzzle is a feature of the Java agent that ensures API compatibility
+between libraries/symbols on the application classpath and APIs of instrumented
+3rd party libraries used by the Agent. In other words the Muzzle ensures
+that the API symbols used by the Agent are compatible with API symbols
+on the application classpath. The Muzzle will prevent loading an instrumentation
+if the APIs do not match.
+
+## How does it work
+
+At build time the Muzzle gradle plugin generates a matcher for 3rd party APIs used by the agent.
+The matcher is stored in the instrumentation class in method `ReferenceMatcher getInstrumentationMuzzle()`.
+
+At runtime the Muzzle checks API compatibility between symbols used by the Agent
+and symbols in the application class loader. Before inspecting the class loader
+the Muzzle plugin checks symbols defined in `SomeInstrumentation.typeMatcher()`
+as a performance optimization. If the symbols do not match the instrumentation
+is not loaded.
+
+## Muzzle gradle plugin
+
+The `printReferences` task prints 3rd party API references in a given module.
+
+```bash
+./gradlew :instrumentation:google-http-client-1.19:printReferences
+```
+
+## Muzzle location
+
+* `buildSrc` - Muzzle Gradle plugin
+* `agent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle` - Muzzle ByteBuddy plugin
```

Comment:
Is this correct? In this location there is `MuzzleGradlePlugin` but it more seems like ByteBuddy plugin as it implements bytebuddy plugin interface.

Reply:
cc) @tylerbenson

Reply:
The ByteBuddy plugin is what does the source graph traversal to find all the references for purposes of matching at run time as well as build time in the `muzzle` task.

Reply:
Yeah, it's a ByteBuddy plugin. The name's a bit misleading and it should be changed in the future -- though to be honest I don't have any great idea right now.

Reply:
We could rename it to `MuzzleByteBuddyPlugin`

Reply:
I created #1288 for renaming it